src/stableswapCalculator.ts

```diff
@@ -7,6 +8,229 @@ const MAX_GET_PRICE_RETRIES = 5;
 
 export class ConvergenceError extends PactSdkError {}
 
+function dAbs(value: bigint) {
+  return value >= 0 ? value : -value;
+}
+
+export function getStableswapMintedLiquidityTokens(
+  addedPrimary: bigint,
+  addedSecondary: bigint,
+  totalPrimary: bigint,
+  totalSecondary: bigint,
+  totalLiquidity: bigint,
+  amplifier: bigint,
+  precision: bigint,
+  feeBps: bigint,
+): bigint {
+  if (totalPrimary + totalSecondary === 0n) {
+    return getConstantProductMintedLiquidityTokens(
+      addedPrimary,
+      addedSecondary,
+      totalPrimary,
+      totalSecondary,
+      totalLiquidity,
+    );
+  }
+
+  const initialTotals: [bigint, bigint] = [totalPrimary, totalSecondary];
+  const updatedTotals: [bigint, bigint] = [
+    totalPrimary + addedPrimary,
+    totalSecondary + addedSecondary,
+  ];
+
+  const [fees, initialD] = getAddLiquidityFees(
+    initialTotals,
+    updatedTotals,
+    feeBps,
+    amplifier,
+    precision,
+  );
+  const nextD = getInvariant(
+    updatedTotals[0] - fees[0],
+    updatedTotals[1] - fees[1],
+    amplifier,
+    precision,
+  );
+
+  return (totalLiquidity * (nextD - initialD)) / initialD;
+}
+
+export function getAddLiquidityBonusPct(
+  addedPrimary: bigint,
+  addedSecondary: bigint,
+  totalPrimary: bigint,
+  totalSecondary: bigint,
+  feeBps: bigint,
+  amplifier: bigint,
+  precision: bigint,
+) {
+  if (totalPrimary + totalSecondary === 0n) {
+    return 0;
+  }
+
+  const initialTotals: [bigint, bigint] = [totalPrimary, totalSecondary];
+  const updatedTotals: [bigint, bigint] = [
+    totalPrimary + addedPrimary,
+    totalSecondary + addedSecondary,
+  ];
+
+  const [fees, initialD] = getAddLiquidityFees(
```

Comment:
I am not 100% sure why `getAddLiquidityFees` returns `initialD`. Do you think it's good idea to pass initialD here as parameter?

Reply:
Of course we can leave it as is, because it's just a small confusion and I don't have good idea how to change it exactly

Reply:
nah, let's leave it.